app/jobs/user_session_jobs/schedule_daily_messages_job.rb

```diff
@@ -0,0 +1,130 @@
+      questions_per_day = question_group.sms_schedule['questions_per_day'] || 1
+      question_group_questions = question_group.questions.order(position: :asc)
+      last_question_index = question_group_questions.last.position
+      first_question_index = question_group_questions.first.position
+      base_range = (first_question_index..last_question_index).to_a
+
+      question_positions = if !last_answer || last_answer.question.position == last_question_index
+                             positions_range = base_range
+                             (questions_per_day / base_range.length.to_f).ceil.to_i.times do
+                               positions_range += base_range
+                             end
+                             positions_range
+                           elsif last_answer.question.position + questions_per_day > last_question_index
+                             from_current_question_range = ((last_answer.question.position + 1)..last_question_index).to_a
+                             ((questions_per_day / base_range.length.to_f).ceil - 1).to_i.times do
+                               from_current_question_range += base_range
+                             end
+                             from_current_question_range
+                           else
+                             ((last_answer.question.position + 1)..last_question_index).to_a
+                           end
+
+      positions_to_be_send = question_positions.first(questions_per_day)
+
+      positions_to_be_send.each_with_index do |position, index|
+        question = question_group.questions.find_by(position: position)
+        time_to_send = calculate_question_sending_time(get_proper_sending_period(@user_session.user_intervention, question_group.sms_schedule),
+                                                       questions_per_day, index)
+        questions_to_be_send_today << { question: question, time_to_send: time_to_send }
+      end
+    end
+
+    # Remove all Information questions if there is any question requiring attention
+    any_answer_expected = questions_to_be_send_today.any? { |elem| elem[:question].type.eql?('Question::Sms') }
+    questions_to_be_send_today.reject! { |elem| elem[:question].type.match('Question::SmsInformation') } if any_answer_expected
+
+    # Remove all questions, that should be send today, but before job execution - case useful on first job scheduling
+    questions_to_be_send_today.reject! { |elem| elem[:time_to_send] < DateTime.current }
+
+    # Schedule all sending jobs
+    questions_to_be_send_today.each do |elem|
+      UserSessionJobs::SendQuestionSmsJob.set(wait_until: elem[:time_to_send])
+                                         .perform_later(@user.id, elem[:question].id, @user_session.id)
+    end
+
+    UserSessionJobs::ScheduleDailyMessagesJob.set(wait_until: DateTime.current.midnight + 1.day)
+                                             .perform_later(user_session_id)
+  end
```

Comment:
What do you think about moving this to a cron job? This is something cyclical, so a cron job should be perfect

Reply:
I thought about it, but I really like the idea of each user session having its own periodic job. It is easy maintainable, trackable and potentially fixable. Having cronjob imo will lead to bigger mess. If you insist on having cronjob, I would propose getting rid of recursively setting new job and schedule all jobs by cronjob. Cronjob will search all active user_sessions and then fire scheduling jobs. Little drawback of this idea - cronjob will schedule `schedule_daily_messages_job` and then `schedule_daily_messages_job ` will schedule `send_question_sms_job`s. For me - overcomplicated.